tidy3d/components/grid/mesher.py

```diff
@@ -63,6 +73,63 @@ class GradedMesher(Mesher):
     """Implements automatic nonuniform meshing with a set minimum steps per wavelength and
     a graded mesh expanding from higher- to lower-resolution regions."""
 
+    def insert_snapping_points(
+        self,
+        axis: Axis,
+        interval_coords: ArrayFloat1D,
+        max_dl_list: ArrayFloat1D,
+        snapping_points: List[Coordinate],
+    ) -> Tuple[ArrayFloat1D, ArrayFloat1D]:
+        """Insert snapping_points to the intervals.
+
+        Parameters
+        ----------
+        axis : Axis
+            Axis index along which to operate.
+        interval_coords : ArrayFloat1D
+            Coordinate of interval boundaries.
+        max_dl_list : ArrayFloat1D
+            Maximal allowed step size of each interval generated from `parse_structures`.
+        snapping_points : List[Coordinate]
+            A set of points that enforce grid boundaries to pass through them.
+
+        Returns
+        -------
+        interval_coords : Array
+            An array of coordinates, where the first element is the simulation min boundary, the
+            last element is the simulation max boundary, and the intermediate coordinates are all
+            locations with a fixpoint or a structure has a bounding box edge along the specified axis.
+            The boundaries are filtered such that no interval is smaller than the smallest
+            of the ``max_steps``.
+        max_steps : array_like
+            An array of size ``interval_coords.size - 1`` giving the maximum grid step required in
+            each ``interval_coords[i]:interval_coords[i+1]`` interval, depending on the materials
+            in that interval, the supplied wavelength, and the minimum required step per wavelength.
+
+        """
+        # Don't do anything for a 2D-like simulation, or no fix points
+        if interval_coords.size == 1 or len(snapping_points) < 1:
+            return interval_coords, max_dl_list
+
+        min_step = np.amin(max_dl_list) * 0.5
+        for point in snapping_points:
+            new_coord = point[axis]
+            # Skip if the point is outside the domain
+            if new_coord >= interval_coords[-1] or new_coord <= interval_coords[0]:
+                continue
+            # search insertion location
+            ind = np.searchsorted(interval_coords, new_coord, side="left")
+            # Skip snapping_points if the distance to the existing interval boundarires are
+            # smaller than half of the minimal maximal step size
+            if abs(new_coord - interval_coords[ind]) < min_step:
+                continue
+            if abs(new_coord - interval_coords[ind - 1]) < min_step:
+                continue
+            # insertion
```

Comment:
yeah, I tend to favor prioritizing structure to avoid shifting interval boundaries when adding a snapping point very close to the structure boundary. In that case, if a snapping point is added inappropriately, at least it doesn't mess up numeric dispersion accuracy.